File: nomi_protocol/events.py
"""Nomi remote vNext SSE event 模型。"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import Field

from nomi_protocol.remote import SSE_EVENT_TYPES
from nomi_protocol.shared import (
    ActiveProviderSelection,
    DesktopSidebarData,
    ProviderStateItem,
    ProviderStateSnapshot,
    SessionMessage,
    SessionSummary,
    StrictBase,
    TaskItem,
)

SseEventType = Literal[*SSE_EVENT_TYPES]


class SseEventEnvelope(StrictBase):
    """SSE 事件统一 envelope。"""

    id: str
    type: SseEventType
    created_at_ms: int
    data: dict[str, Any] = Field(default_factory=dict)


class RuntimeConnectedData(StrictBase):
    """runtime.connected 数据。"""

    version: str
    provider_state: ProviderStateSnapshot


class RuntimeStatusChangedData(StrictBase):
    """runtime.status_changed 数据。"""

    status: dict[str, Any] = Field(default_factory=dict)


class RuntimeReloadedData(StrictBase):
    """runtime.reloaded 数据。"""

    active: ActiveProviderSelection
    provider_state: ProviderStateSnapshot


class SessionData(StrictBase):
    """session created/updated/deleted 数据。"""

    session: SessionSummary


class SessionMessageAppendedData(StrictBase):
    """session.message_appended 数据。"""

    session_id: str
    message: SessionMessage
    index: int | None = None


class TurnStartedData(StrictBase):
    """turn.started 数据。"""

    turn_id: str
    session_id: str


class TurnTextData(StrictBase):
    """turn progress/delta 数据。"""

    turn_id: str | None = None
    session_id: str
    content: str
    tool_hint: bool = False


class TurnStreamEndData(StrictBase):
    """turn.stream_end 数据。"""

    turn_id: str | None = None
    session_id: str
    resuming: bool = False


class TurnCompletedData(StrictBase):
    """turn completed/failed/interrupted 数据。"""

    turn_id: str | None = None
    session_id: str
    stop_reason: str
    error: str | None = None


class TaskEventData(StrictBase):
    """task 事件数据。"""

    task: TaskItem | None = None
    task_id: str | None = None
    content: str | None = None
    session_id: str | None = None


class ProviderStateChangedData(StrictBase):
    """provider 状态事件数据。"""

    provider_state: ProviderStateSnapshot


class ProviderSettingsUpdatedData(StrictBase):
    """provider 设置事件数据。"""

    provider: str
    settings: ProviderStateItem
    requires_runtime_reload: bool


class ProviderActiveChangedData(StrictBase):
    """active provider 事件数据。"""

    active: ActiveProviderSelection
    requires_runtime_reload: bool


class SidebarSnapshotData(StrictBase):
    """sidebar.snapshot 数据。"""

    sidebar: DesktopSidebarData


class ResourceEventData(StrictBase):
    """skill/mcp 资源事件数据。"""

    resource: str
    action: str
    name: str | None = None
    item: dict[str, Any] | None = None

File: src/events.ts
import { SSE_EVENT_TYPES } from "./spec.js";
import type {
  ActiveProviderSelection,
  DesktopSidebarData,
  ProviderStateItem,
  ProviderStateSnapshot,
  SessionMessage,
  SessionSummary,
  TaskItem,
} from "./shared.js";

export { SSE_EVENT_TYPES };
export type SseEventType = (typeof SSE_EVENT_TYPES)[number];

export interface SseEventEnvelope<TData extends Record<string, unknown> = Record<string, unknown>> {
  id: string;
  type: SseEventType;
  created_at_ms: number;
  data: TData;
}

export interface RuntimeConnectedData {
  version: string;
  provider_state: ProviderStateSnapshot;
}

export interface RuntimeStatusChangedData {
  status: Record<string, unknown>;
}

export interface RuntimeReloadedData {
  active: ActiveProviderSelection;
  provider_state: ProviderStateSnapshot;
}

export interface SessionData {
  session: SessionSummary;
}

export interface SessionMessageAppendedData {
  session_id: string;
  message: SessionMessage;
  index?: number | null;
}

export interface TurnStartedData {
  turn_id: string;
  session_id: string;
}

export interface TurnTextData {
  turn_id?: string | null;
  session_id: string;
  content: string;
  tool_hint: boolean;
}

export interface TurnStreamEndData {
  turn_id?: string | null;
  session_id: string;
  resuming: boolean;
}

export interface TurnCompletedData {
  turn_id?: string | null;
  session_id: string;
  stop_reason: string;
  error?: string | null;
}

export interface TaskEventData {
  task?: TaskItem | null;
  task_id?: string | null;
  content?: string | null;
  session_id?: string | null;
}

export interface ProviderStateChangedData {
  provider_state: ProviderStateSnapshot;
}

export interface ProviderSettingsUpdatedData {
  provider: string;
  settings: ProviderStateItem;
  requires_runtime_reload: boolean;
}

export interface ProviderActiveChangedData {
  active: ActiveProviderSelection;
  requires_runtime_reload: boolean;
}

export interface SidebarSnapshotData {
  sidebar: DesktopSidebarData;
}

export interface ResourceEventData {
  resource: string;
  action: string;
  name?: string | null;
  item?: Record<string, unknown> | null;
}
